File: ReactNativeApp/ios/Modules/Phases/Games/Instance/JugglingPhase.swift
//
//  JugglingPhase.swift
//  jogo
//
//  Created by arham on 19/11/2021.
//

import Foundation

public class JugglingPhase: GamePhase {
  
  var game: SSVC<JugglingState, JugglingView, JugglingController>?
  
  required init(delegate: GamePhaseDelegate, stopRestart: Bool) {
    super.init(delegate: delegate, stopRestart: stopRestart)
  }
  
  public override func initialize() {
    self.phaseName = "Juggling"
    game = JugglingBuilder.build(ballController: ball?.controller,
                                 personController: person?.controller,
                                 pointController: points?.controller)
    super.initialize()
    Profiler.get().profileTime(frameId: GameContext.getContext()?.currentFrameNumber ?? -1, tag: "\(self.phaseName ?? "" )-\(#function)", delta: Date().getMilliseconds() - timings["\(#function)"]!)
  }
  
  public override func process() {
    super.process()
    do {
      guard let context = GameContext.getContext() else { return }
      let _infoBlob: InfoBlob? = try context.infoBlobArrayList.getLast()
      guard let infoBlob = _infoBlob else { return }
      game?.controller.process(infoBlob: infoBlob)
    } catch {
      Logger.shared.logError(text: "Context InfoBlob last value was not retreivable.")
    }
    Profiler.get().profileTime(frameId: GameContext.getContext()?.currentFrameNumber ?? -1, tag: "\(self.phaseName ?? "" )-\(#function)", delta: Date().getMilliseconds() - timings["\(#function)"]!)
  }
  
  public override func cleanup() {
    self.game?.cleanup()
    self.game = nil
    super.cleanup()
  }
  
  public override func getDebugText() -> String {
    return ""
  }
}

File: ReactNativeApp/android/app/src/main/java/com/phases/games/instances/JugglingPhase.java
package com.phases.games.instances;

import com.components.games.base.GameContext;
import com.components.games.instances.juggling.JugglingBuilder;
import com.components.games.instances.juggling.JugglingController;
import com.components.games.instances.juggling.JugglingState;
import com.components.games.instances.juggling.JugglingView;
import com.phases.instances.GamePhase;
import com.ssvc.SSVC;

public class JugglingPhase extends GamePhase {

    SSVC<JugglingState, JugglingView, JugglingController> jugglingGame;


    @Override
    public void initialize() {
        super.initialize();
        this.jugglingGame = JugglingBuilder.buildJugglingGame(points.controller,streak.controller);

    }

    @Override
    public void process() {
        super.process();
        if (jugglingGame == null) return;
        jugglingGame.controller.process();
    }

    @Override
    public void cleanup() {
        jugglingGame.stateView.detachFromLayoutSV();
        GameContext.getContext().setRelevantGameInfo("score", String.valueOf(points.state.currentPoints));
        jugglingGame = null;
        super.cleanup();
    }
}
